core/util/authentication_for_opds.py

```diff
@@ -74,7 +76,9 @@ def to_dict(self, _db):
         document = dict(id=self.id, title=self.title)
         flow_documents = document.setdefault("authentication", [])
         for flow in self.authentication_flows:
-            flow_documents.append(flow.authentication_flow_document(_db))
+            doc = flow.authentication_flow_document(_db)
+            if doc is not None:
+                flow_documents.append(flow.authentication_flow_document(_db))
```

Comment:
This allows us to skip flows in the authentication document

Reply:
Why do we need to skip flows?

Reply:
I'm kind of unsure about this. The clients used basic authentication to retrieve a token, that part I understand. The problem is that (given that I know a bit about how the Android client is implemented), they're going to need to understand that this "basic authentication" behaves differently to the other kind of "basic authentication" they'll see elsewhere, and that they'll need to store and use a token repeatedly. I feel like this information _should_ be in the authentication document.

Reply:
I don't think there is any documented protocol to define that in the authentication document. I think the prevailing thought here was that this would be written in directly into the code of the Basic Auth implementations and not controlled via the authentication document.

Reply:
Eg.
```
"authentication": [
    {
        "description": "Library Barcode",
        "labels": {
            "login": "Barcode",
            "password": "PIN"
        },
        "inputs": {
            "login": {
                "keyboard": "Default"
            },
            "password": {
                "keyboard": "Default"
            }
        },
        "type": "http://opds-spec.org/auth/basic",
        "secondary": {
           "type": "http://opds-spec.org/auth/bearer",
           "href": ".../patrons/me/token",
        }
    }
],
```

Reply:
I feel like it should be treated as an entirely separate authentication method, with a different `type`.

Doing so unambiguously signals to the client that they can use a different login workflow to use tokens (and by putting the token option in the authentication document ahead of plain basic authentication, it signals to them that they _should_ use tokens in preference to plain basic authentication). The clients already have separate code paths for basic auth, OAUTH, and SAML, so adding another authentication method shouldn't be a problem.

It also allows us to turn off plain basic authentication eventually, when we've determined that no clients have used it for _insert time period here_.

Reply:
As an aside, the Android client is using sealed types for the credentials, so if @nunommts adds a new `PalaceBearerToken` credentials type, he'll get nice errors indicating all of the places in the codebase that now need updating. :slightly_smiling_face: